src/library/blast/arith/heuristic.cpp

```diff
@@ -0,0 +1,552 @@
+#include "library/blast/simplifier/simplifier.h"
+#include "library/num.h"
+
+#ifndef LEAN_DEFAULT_HEURISTIC_MAX_STEPS_PER_ACTION
+#define LEAN_DEFAULT_HEURISTIC_MAX_STEPS_PER_ACTION 10
+#endif
+
+namespace lean {
+namespace blast {
+
+/** Globals */
+static name * g_heuristic_trace_name           = nullptr;
+static name * g_heuristic_max_steps_per_action = nullptr;
+
+static unsigned g_ext_id = 0;
+
+struct arith_heuristic_proof_step_cell : public proof_step_cell {
+    list<expr>     m_placeholders;     // placeholders for non-zero proofs
+
+    virtual ~arith_heuristic_proof_step_cell() {}
+    arith_heuristic_proof_step_cell(list<expr> const & placeholders):
+        m_placeholders(placeholders) {}
+
+    virtual action_result resolve(expr const & pf) const override {
+        buffer<expr> proofs;
+        for (expr const & placeholder : m_placeholders) {
+            optional<expr> pf;
+            {
+                flet<state> save_state(curr_state(), curr_state());
+                curr_state().set_target(mlocal_type(placeholder));
+                pf = mk_backward_strategy()();
+            }
+            if (pf) {
+                proofs.push_back(*pf);
+            } else {
+                lean_trace(*g_heuristic_trace_name, tout() << "failed to prove: " << ppb(mlocal_type(placeholder)) << "\n";);
+                return action_result::failed();
+            }
+        }
+        buffer<expr> placeholders;
+        to_buffer(m_placeholders, placeholders);
+        return action_result::solved(mk_app(Fun(placeholders, pf), proofs));
+    }
+};
+
+void addmul(mpq & q, mpq const & c, mpq const & x) {
+    mpq tmp = c;
+    tmp *= x;
+    q += tmp;
+}
+
+/** Option getters */
+unsigned get_heuristic_max_steps_per_action() {
+    return ios().get_options().get_unsigned(*g_heuristic_max_steps_per_action, LEAN_DEFAULT_HEURISTIC_MAX_STEPS_PER_ACTION);
+}
+
+/* Description:
+
+This module performs linear and some non-linear arithmetic, as described in the paper [HEURISTIC2].
+
+The architecture is as follows:
+
+When the action [assert_heuristic_action(hypothesis_idx)] is called, we
+call the [linearizer] to construct a list of polynomial inequalities, of the form
+
+<<
+0 < Sum_i [<numeral_i> * <unknown_i>] + <numeral>
+>>
+
+where the inequality may or may not be strict.
+
+The linearizer stores a [lazy_proof] in each polynomial inequality that it creates, so that
+we can construct a fully formal proof of any contradiction we find.
+
+Then, we proceed to create new polynomial inequalities from old polynomial inequaities, all the while
+maintaining a trace of how the new polynomial inequalities can be justified in terms of the old ones.
+
+If we do construct a polynomial inequality of the form [0 < 0, 0 < - <pos>, 0 <= - <pos>], we have
+found a contradiction, and we must start to construct a formal proof of this fact.
+*/
+
+/** Basic data-structures **/
+
+/* Proof trails */
+typedef std::function<expr()> lazy_proof;
```

Comment:
It is not even helpful now, since we always prove the original un-normalized `0 < ...` eagerly. However, if the linearizer is extended this flexibility may prove useful. I can get rid of it for now and just store the proof itself.